src/function/scalar/nested/list_extract.cpp

```diff
@@ -124,23 +122,66 @@ static void ListExtractFunFun(DataChunk &args, ExpressionState &state, Vector &r
 		throw NotImplementedException("Unimplemented type for LIST_EXTRACT");
 	}
 
-	result.Verify(args.size());
+	result.Verify(count);
+}
+
+static void StringExtractFunction(Vector &result, Vector &input_vector, Vector &subscript_vector, const idx_t count) {
+	BinaryExecutor::Execute<string_t, int32_t, string_t>(
+	    input_vector, subscript_vector, result, count, [&](string_t input_string, int32_t subscript) {
+		    return SubstringFun::SubstringScalarFunction(result, input_string, subscript + int32_t(subscript >= 0), 1);
+	    });
 }
 
-static unique_ptr<FunctionData> ListExtractBind(ClientContext &context, ScalarFunction &bound_function,
-                                                vector<unique_ptr<Expression>> &arguments) {
-	if (arguments[0]->return_type.id() != LogicalTypeId::LIST) {
-		throw BinderException("LIST_EXTRACT can only operate on LISTs");
+static void ArrayExtractFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	D_ASSERT(args.ColumnCount() == 2);
+	auto count = args.size();
+
+	result.SetVectorType(VectorType::CONSTANT_VECTOR);
+	for (idx_t i = 0; i < args.ColumnCount(); i++) {
+		if (args.data[i].GetVectorType() != VectorType::CONSTANT_VECTOR) {
+			result.SetVectorType(VectorType::FLAT_VECTOR);
+		}
+	}
+
+	Vector &base = args.data[0];
+	Vector &subscript = args.data[1];
+
+	switch (base.GetType().id()) {
+	case LogicalTypeId::LIST:
+		// Share the value dictionary as we are just going to slice it
+		ListExtractFunction(result, base, subscript, count);
+		break;
+	case LogicalTypeId::VARCHAR:
+		StringExtractFunction(result, base, subscript, count);
+		break;
+	default:
+		throw NotImplementedException("Specifier type not implemented");
+	}
+}
+
+static unique_ptr<FunctionData> ArrayExtractBind(ClientContext &context, ScalarFunction &bound_function,
+                                                 vector<unique_ptr<Expression>> &arguments) {
+	D_ASSERT(bound_function.arguments.size() == 2);
+	switch (arguments[0]->return_type.id()) {
+	case LogicalTypeId::LIST:
+		// list extract returns the child type of the list as return type
+		bound_function.return_type = arguments[0]->return_type.child_types()[0].second;
+		break;
+	case LogicalTypeId::VARCHAR:
+		// string extract returns a string, but can only accept 32 bit integers
+		bound_function.return_type = arguments[0]->return_type;
```

Comment:
Rather than manually binding on the first type here, could you use the function binder by creating function overloads for LIST and VARCHAR, or does that not work with LIST arguments? Perhaps the binder needs to be adapted to support `LIST<ANY>` as an input parameter.

Reply:
I copied it from existing code, but I'll see if I can make this work.

Reply:
Yes it does. `CastRules::ImplicitCast` doesn't seem to understand `LIST` types.